server/entity/splash_potion.go

```diff
@@ -0,0 +1,245 @@
+	"github.com/df-mc/dragonfly/server/entity/physics/trace"
+	"github.com/df-mc/dragonfly/server/internal/nbtconv"
+	"github.com/df-mc/dragonfly/server/item/potion"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/df-mc/dragonfly/server/world/particle"
+	"github.com/df-mc/dragonfly/server/world/sound"
+	"github.com/go-gl/mathgl/mgl64"
+	"image/color"
+	"math"
+	"time"
+)
+
+// SplashPotion is an item that grants effects when thrown.
+type SplashPotion struct {
+	transform
+	yaw, pitch float64
+
+	ticksLived int
+
+	closeNextTick bool
+
+	owner world.Entity
+
+	variant potion.Potion
+
+	c *ProjectileComputer
+}
+
+// NewSplashPotion ...
+func NewSplashPotion(pos mgl64.Vec3, yaw, pitch float64, owner world.Entity) *SplashPotion {
+	s := &SplashPotion{
+		yaw:   yaw,
+		pitch: pitch,
+		c: &ProjectileComputer{&MovementComputer{
+			Gravity:           0.05,
+			Drag:              0.01,
+			DragBeforeGravity: true,
+		}},
+		owner: owner,
+	}
+	s.transform = newTransform(s, pos)
+
+	return s
+}
+
+// Name ...
+func (s *SplashPotion) Name() string {
+	return "Splash Variant"
+}
+
+// EncodeEntity ...
+func (s *SplashPotion) EncodeEntity() string {
+	return "minecraft:splash_potion"
+}
+
+// AABB ...
+func (s *SplashPotion) AABB() physics.AABB {
+	return physics.NewAABB(mgl64.Vec3{-0.125, 0, -0.125}, mgl64.Vec3{0.125, 0.25, 0.125})
+}
+
+// SetVariant ...
+func (s *SplashPotion) SetVariant(variant potion.Potion) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.variant = variant
+}
+
+// Variant ...
+func (s *SplashPotion) Variant() potion.Potion {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.variant
+}
+
+// Rotation ...
+func (s *SplashPotion) Rotation() (float64, float64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.yaw, s.pitch
+}
+
+// splashable represents an entity that can be splashed by a potion.
+type splashable interface {
+	Living
+	Eyed
+
+	// AddEffect adds a specific effect to the entity that implements this interface.
+	AddEffect(e effect.Effect)
+}
+
+// Tick ...
+func (s *SplashPotion) Tick(current int64) {
+	if s.closeNextTick {
+		_ = s.Close()
+		return
+	}
+	s.mu.Lock()
+	m, result := s.c.TickMovement(s, s.pos, s.vel, s.yaw, s.pitch, s.ignores)
+	s.pos, s.vel, s.yaw, s.pitch = m.pos, m.vel, m.yaw, m.pitch
+	s.mu.Unlock()
+
+	s.ticksLived++
+	m.Send()
+
+	if m.pos[1] < cube.MinY && current%10 == 0 {
+		s.closeNextTick = true
+		return
+	}
+
+	if result != nil {
+		w := s.World()
+		pos := s.Position()
+		aabb := s.AABB().Translate(pos)
+
+		pot := s.variant
+		effects := pot.Effects
+		hasEffects := len(effects) > 0
+
+		colour := color.RGBA{R: 0x38, G: 0x5d, B: 0xc6, A: 0xff}
+		if hasEffects {
+			colour, _ = effect.ResultingColour(effects)
+		}
+
+		w.AddParticle(pos, particle.Splash{Colour: colour})
+		w.PlaySound(pos, sound.GlassBreak{})
+
+		if hasEffects {
+			ignoreFunc := func(entity world.Entity) bool {
+				_, canSplash := entity.(splashable)
+				if !canSplash || entity == s {
```

Comment:
You can get rid of this if statement and just have `return !canSplash || entity == s`